server/lib/notifications/notification-service.ts

```diff
@@ -0,0 +1,71 @@
+import { NydusServer } from 'nydus'
+import { singleton } from 'tsyringe'
+import {
+  NotificationAddEvent,
+  NotificationClearByIdEvent,
+  NotificationServerInitEvent,
+} from '../../../common/notifications'
+import { ClientSocketsManager } from '../websockets/socket-groups'
+import {
+  addNotification,
+  clearById,
+  NotificationData,
+  retrieveNotifications,
+} from './notification-model'
+
+export function getNotificationsPath(userId: number): string {
+  return `/notifications/${userId}`
+}
+
+@singleton()
+export default class NotificationService {
+  constructor(private nydus: NydusServer, private clientSocketsManager: ClientSocketsManager) {
+    this.clientSocketsManager.on('newClient', async c => {
```

Comment:
Event handlers are not expected to be `async`, so this is adding a new spot that has potential rejected promises that get dropped on the floor. If you'd like to use an async function, you need to attach an error handler to it and do something with those errors. Likely in this case, it's a lot simpler to just write

```ts
  retrieveNotifications(...).then(notifications => {
  
  }, err => {

  })
```

Reply:
I'll just note that this is also a race condition, if something sends this user a notification between the time this query hits the DB and comes back, they will not see that notification from this message *or* any message sent to the route (it won't have been subscribed yet).

Fixing this is probably a bit of a pain:

  - You could add promise locks around initialization that are awaited in mutations
  - You could keep track of when clients are being initialized and store info about the mutations that occurred during that, such that it could be appended to the `serverInit` message or just simply re-sent after (probably fine to re-send given IDs are unique)
  - Something else :)

(would be really nice if the nydus setup let us us subscribe and then send initial data in some async timeframe afterwards =/ )

Reply:
What about if we simply subscribe the user to the notifications path immediately when they connect, and then just publish them the server-init notifications once they're fetched from the DB?

Reply:
Pretty wasteful, because you're publishing to every connected client for that user then. You could subscribe the user to two paths, e.g. `/notifications/<user>/` and `/notifications/<user>/<clientId>/` to make that work (still not perfectly, since they might have multiple connections for the same client at some points, but it's at least a bit better).

I think I may just try to make nydus allow for this though, doesn't seem like that huge of a change